### PR TITLE
[spirv] Plumb NCHW convolution support through

### PR DESCRIPTION
**Pull request by antiagainst**
This commit enables tiling and vectorizing NCHW convolution ops.
The vectorization flow still needs further improvements, like
removing `vector.extract`/`vector.insert` pairs due to `vector.transpose`,
enabling hoisting `vector.transfer_*` ops, etc. This commit should give
a basis.

Towards https://github.com/iree-org/iree/issues/10536

### CODE REVIEW

**Review by ThomasRaoux — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/SPIRV/KernelConfig.cpp [hunk 1]
```diff
@@ -131,74 +131,101 @@ LogicalResult setConvOpConfig(linalg::LinalgOp linalgOp,
...
-
-  // Otherwise treat OW and OH separately to allow them to have different number
-  // of threads and tiling size.
-  if (!tileToSquare) {
-    if (!distributeToOneDim(ow, /*isInnerMostDim=*/false, residualThreads,
+  SmallVector<int64_t, 3> workgroupSize(3, 1);  // (X, Y, Z)
+  SmallVector<int64_t> workgroupTileSizes(4, 0);
+  SmallVector<int64_t> threadTileSizes(4, 0);
+
+  if (isNCHW) {
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/SPIRV/KernelConfig.cpp:168]: nit: would it be possible to split this function into a couple smaller ones? I don't have any concrete issues with it, but it's getting quite long

Comment (antiagainst) [reply compiler/src/iree/compiler/Codegen/SPIRV/KernelConfig.cpp:168]: I did a refactoring prior to this to extract quite some logic out https://github.com/iree-org/iree/pull/10571. Further splitting would mean duplicating quite some common code. So I'd keep it for now.

File: compiler/src/iree/compiler/Codegen/SPIRV/KernelConfig.cpp [hunk 2]
```diff
@@ -131,74 +131,101 @@ LogicalResult setConvOpConfig(linalg::LinalgOp linalgOp,
...
   ArrayRef<int64_t> outputShape = outputType.cast<ShapedType>().getShape();
 
-  if (isa<linalg::Conv2DNhwcHwcfOp>(*linalgOp) &&
-      ShapedType::isDynamic(inputShape[3])) {
-    return success();
-  }
-  if (llvm::any_of(outputShape.drop_front(), ShapedType::isDynamic)) {
+  const bool isNCHW = isa<linalg::Conv2DNchwFchwOp>(*linalgOp);
+  const bool isNHWC = isa<linalg::Conv2DNhwcHwcfOp>(*linalgOp) ||
+                      isa<linalg::DepthwiseConv2DNhwcHwcOp>(*linalgOp);
```
[the remark this answers is not in the document]
Comment (antiagainst) [reply compiler/src/iree/compiler/Codegen/SPIRV/KernelConfig.cpp:136]: To me this formats better. :)

**Comment by antiagainst** [comment]
`test_shark_model_suite` is failing due to dependency installation issue: "Installing build dependencies: finished with status 'error'". That's unrelated to this one. So I'll ignore it and land this patch.

**Pull request merged**